modules/internal/ChapelIO.chpl

```diff
@@ -345,50 +349,225 @@ module ChapelIO {
+                 isParam(__primitive("field value by num", x, i)) {
+                // do nothing, don't read types or params
               } else {
-                eq = new ioLiteral("=", true);
+
+                if !read_field_name && !read_field[i] {
+                  if st == QIO_AGGREGATE_FORMAT_JSON {
+                    fname = new ioLiteral('"' +
+                                          __primitive("field num to name", t, i)
+                                          + '"',
+                                          true);
+                  } else {
+                    fname = new ioLiteral(__primitive("field num to name", t, i),
+                                          true);
+                  }
+
+                  //writeln("ERROR IS ", reader.error():int);
+                  //writeln("OFFSET IS ", reader.offset());
+                  //writeln("TRYING TO READ ", fname);
+                  reader.readwrite(fname);
+                  //writeln("POST FNAME ERROR IS ", reader.error():int);
+
+                  if reader.error() == EFORMAT || reader.error() == EEOF {
+                    // Try reading again with a different union element.
+                    reader.clearError();
+                  } else {
+                    read_field_name = true;
+                    needsComma = true;
+
+                    //writeln("DID     FIND ", fname);
+
+                    var eq:ioLiteral;
+                    if st == QIO_AGGREGATE_FORMAT_JSON {
+                      eq = new ioLiteral(":", true);
+                    } else {
+                      eq = new ioLiteral("=", true);
+                    }
+                    reader.readwrite(eq);
+
+                    reader.readwrite(__primitive("field value by num", x, i));
+                    if !reader.error() {
+                      //writeln("setting read_field[", i, "]");
+                      read_field[i] = true;
+                      num_read += 1;
+                    }
+                  }
+                }
               }
-              reader.readwrite(eq);
             }
-    
-            reader.readwrite(__primitive("field value by num", x, i));
-    
-            first = false;
+
+            //writeln("AFTER PARAM FOR");
+
+            // Stop with an error if we didn't read a field name
+            // ... unless we skip unknown fields...
+            if !read_field_name {
+              if skip_unk != 0 && st == QIO_AGGREGATE_FORMAT_JSON {
+
+                //writeln("SKIPPING FIELD");
+
+                // Skip an unknown JSON field.
+                var e:syserr;
+                reader.skipField(error=e);
+                if !e {
+                  needsComma = true;
+                }
+                reader.setError(e);
+              } else {
+                //writeln("ERROR - MISSING FIELD NAME");
+                reader.setError(EFORMAT:syserr);
+                break;
+              }
```

Comment:
I actually used to have these two cases folded together, but I became much less confused about what the outer loop does when I took them apart. They have slightly different error handling. This chunk of code is short enough that it doesn't seem worth it to me to make a function that behaves differently based on an argument.